File: bindings/python/examples/mcp_integration.py
"""MCP integration: expose SCP context outlets via MCP server, connect as client.

Phase 4 PR 5 (#1549) moved MCP operations onto :class:`scp_sdk.SCP`.
Use :meth:`SCP.mcp_serve`, :meth:`SCP.mcp_client_connect_sse`,
:meth:`SCP.mcp_client_list_tools`, :meth:`SCP.mcp_client_invoke`,
:meth:`SCP.mcp_client_disconnect`, and :meth:`SCP.mcp_server_stop`.
"""

import asyncio

from scp_sdk import SCP
from scp_sdk.types import Capability, CustodyType, MemoryScope


async def main() -> None:
    with SCP(storage={"type": "in_memory"}) as scp:
        identity = await scp.identity_create(CustodyType.IN_MEMORY)

        # Create a context with outlet capabilities.
        ctx = await scp.context_create(
            identity.did,
            {
                "ceiling": [
                    Capability.MESSAGES_READ.value,
                    Capability.MESSAGES_WRITE.value,
                    Capability.OUTLET_CALL_ALL.value,
                    Capability.OUTLET_REGISTER.value,
                ],
                "memory_scope": MemoryScope.EPHEMERAL.value,
                "governance": "single_admin",
            },
        )

        # Start an MCP server exposing context outlets on stdio.
        server = await scp.mcp_serve(identity.did, [ctx.context_id], "stdio")
        print("MCP server running")

        # Or connect as an MCP client to an external server via SSE.
        client = await scp.mcp_client_connect_sse("http://localhost:8080/mcp")
        outlets = await scp.mcp_client_list_tools(client)
        print(f"Remote server offers {len(outlets)} outlet(s)")

        result = await scp.mcp_client_invoke(
            client,
            "summarize",
            {"text": "SCP is a protocol for..."},
            ctx.context_id,
            identity.did,
        )
        print(f"Result: {result}")

        await scp.mcp_client_disconnect(client)
        await scp.mcp_server_stop(server)
        await scp.context_close(ctx._raw_handle, identity.did)


if __name__ == "__main__":
    asyncio.run(main())

File: bindings/typescript/examples/mcp-integration.ts
/**
 * MCP integration: expose SCP outlets via an MCP JSON-RPC server, then
 * connect as an MCP client to a remote server.
 *
 * Post-Phase-4 (ADR-048): all bridge operations route through an
 * explicit `SCP` instance. The free-function shims (`serveMcp`,
 * `connectMcp`, `connectMcpStdio`) were removed — use
 * `scp.mcpServerCreate`, `scp.mcpClientConnectSse`, and the other
 * `scp.mcp*` methods directly.
 *
 * Run: bun run examples/mcp-integration.ts
 */

import { SCP, defineOutletDefinition } from "../src/index";

async function main(): Promise<void> {
  const scp = new SCP({ storage: { type: "in_memory" } });
  try {
    const identity = await scp.identityCreate("in_memory");

    // Create a context with outlet capabilities.
    const ctx = await scp.contextCreate(
      identity,
      JSON.stringify({
        ceiling: ["messages:read", "messages:write", "tool:invoke:*", "tool:register"],
        memoryScope: "ephemeral",
        governance: "single_admin",
      }),
    );

    // Register an outlet in the context.
    const outlet = defineOutletDefinition({
      name: "summarize",
      description: "Summarize text content",
      inputSchema: {
        type: "object",
        properties: { text: { type: "string" } },
        required: ["text"],
      },
      outputSchema: {
        type: "object",
        properties: { summary: { type: "string" } },
      },
      operator: identity.did,
    });
    await scp.outletRegister(ctx._rawHandle, outlet);

    // Start an MCP server exposing context outlets on stdio.
    const server = await scp.mcpServerCreate({
      identityDid: identity.did,
      contextIds: [ctx.contextId],
      transport: "stdio",
    });
    console.log("MCP server running, exposing outlets");

    try {
      // Or connect as an MCP client to a remote server.
      const client = await scp.mcpClientConnectSse("http://localhost:8080/mcp");
      try {
        const outlets = await scp.mcpClientListTools(client);
        console.log(`Remote server offers ${outlets.length} outlet(s)`);

        const result = await scp.mcpClientInvoke(
          client,
          "summarize",
          JSON.stringify({ text: "SCP is a protocol for..." }),
          ctx.contextId,
          identity.did,
        );
        console.log("Result:", result);
      } finally {
        await scp.mcpClientDisconnect(client);
      }
    } finally {
      await scp.mcpServerStop(server);
    }

    await scp.contextClose(ctx._rawHandle, identity.did);
  } finally {
    await scp.shutdown(5);
  }
}

main().catch((error: unknown) => {
  console.error("Demo failed:", error);
  process.exit(1);
});
